pkg/storegateway/chunkscache/cache.go

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package chunkscache
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gokit_log "github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cache"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+// Range is a contiguous range of chunks. Start is the ref of the first chunk and NumChunks is how many chunks
+// there are in the range.
+type Range struct {
+	BlockID   ulid.ULID
+	Start     chunks.ChunkRef
+	NumChunks int
+}
+
+type ChunksCache interface {
+	FetchMultiChunks(ctx context.Context, userID string, ranges []Range) (hits map[Range][]byte)
+	StoreChunks(ctx context.Context, userID string, r Range, v []byte)
+}
+
+type TracingCache struct {
+	c ChunksCache
+	l gokit_log.Logger
+}
+
+func NewTracingCache(c ChunksCache, l gokit_log.Logger) TracingCache {
+	return TracingCache{
+		c: c,
+		l: l,
+	}
+}
+
+func (c TracingCache) FetchMultiChunks(ctx context.Context, userID string, ranges []Range) (hits map[Range][]byte) {
+	hits = c.c.FetchMultiChunks(ctx, userID, ranges)
+
+	l := spanlogger.FromContext(ctx, c.l)
+	level.Debug(l).Log(
+		"name", "ChunksCache.FetchMultiChunks",
+		"ranges_requested", len(ranges),
+		"ranges_hit", len(hits),
+		"ranges_hit_bytes", hitsSize(hits),
+		"ranges_misses", len(ranges)-len(hits),
+	)
+	return
+}
+
+func hitsSize(hits map[Range][]byte) (size int) {
+	for _, b := range hits {
+		size += len(b)
+	}
+	return
+}
+
+func (c TracingCache) StoreChunks(ctx context.Context, userID string, r Range, v []byte) {
+	c.c.StoreChunks(ctx, userID, r, v)
+}
+
+const (
+	defaultTTL = 7 * 24 * time.Hour
+)
+
+// DskitChunksCache is a ChunksCache which uses the dskit caching interface.
+type DskitChunksCache struct {
```

Comment:
I strongly recommend to not name it "dskit" but just `ChunksCache`. We should rename the `type ChunksCache interface` to something different, maybe just `Cache` given the package is already named `chunkscache`?

Reply:
I went with `type ChunksCache struct` and `type Cache interface` in c86b7c99ae87fc2c3aa3a2aac8633a3a5b3b7be2